src/Build/BackEnd/Components/FileAccesses/FileAccessManager.cs

```diff
@@ -0,0 +1,135 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System;
+using System.Collections.Generic;
+using System.Threading;
+using Microsoft.Build.BackEnd;
+using Microsoft.Build.Framework.FileAccess;
+using Microsoft.Build.Shared;
+
+namespace Microsoft.Build.FileAccesses
+{
+    internal sealed class FileAccessManager : IFileAccessManager, IBuildComponent
+    {
+        private IScheduler? _scheduler;
+        private IConfigCache? _configCache;
+
+        private readonly ReaderWriterLockSlim _handlersLock = new();
+        private List<Action<BuildRequest, FileAccessData>> _fileAccessHandlers = new();
+        private List<Action<BuildRequest, ProcessData>> _processHandlers = new();
+
+        public static IBuildComponent CreateComponent(BuildComponentType type)
+        {
+            ErrorUtilities.VerifyThrowArgumentOutOfRange(type == BuildComponentType.FileAccessManager, nameof(type));
+            return new FileAccessManager();
+        }
+
+        public void InitializeComponent(IBuildComponentHost host)
+        {
+            _scheduler = host.GetComponent(BuildComponentType.Scheduler) as IScheduler;
+            _configCache = host.GetComponent(BuildComponentType.ConfigCache) as IConfigCache;
+        }
+
+        public void ShutdownComponent()
+        {
+            _scheduler = null;
+            _configCache = null;
+        }
+
+        public void ReportFileAccess(FileAccessData fileAccessData, int nodeId)
+        {
+            BuildRequest? buildRequest = GetBuildRequest(nodeId);
+            if (buildRequest != null)
+            {
+                _handlersLock.EnterReadLock();
+                try
+                {
+                    foreach (Action<BuildRequest, FileAccessData> handler in _fileAccessHandlers)
+                    {
+                        handler.Invoke(buildRequest, fileAccessData);
+                    }
+                }
+                finally
+                {
+                    _handlersLock.ExitReadLock();
+                }
+            }
+        }
+
+        public void ReportProcess(ProcessData processData, int nodeId)
```

Comment:
Somewhat for logging, and somewhat for "extra validation" that a plugin could do.

For example in QuickBuild, we have an interface-aware caching feature (basically ref assemblies), which validates that it can classify every write, and part of that requires knowing which process did the writes. Eg we expect csc to write the dll, but then if we see other non-copy writes to that dll from other processes, we know it's not necessarily a safe output to replay since we have no idea what the build is doing to the file.